datasets/dataset_plugin.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import annotations
+
+from abc import ABC
+from typing import Callable, Dict, Iterable, Optional, Union
+
+from datasets.context import Context
+
+from .mode import Mode
+from .program_executor import ProgramExecutor
+
+
+class DatasetPlugin(ABC):
+    """
+    All dataset plugins derive from this class.
+    To register as a dataset they must decorate themselves with or call DatasetPlugin.register_plugin()
+    """
+
+    _executor: ProgramExecutor
+    # Context -> constructor_keys -> dataset plugin
+    _plugins: Dict[Context, Dict[set[str], DatasetPlugin]] = {}
+    _META_COLUMNS = ["run_id"]
+
+    def __init__(
+        self,
+        name: str,
+        logical_key: Optional[str] = None,
+        columns: Optional[Union[Iterable[str], str]] = None,
```

Comment:
I'd recommend making the `columns` type into an alias and adding it to the public interface for the library. That way clients can subclass it or add it as a type hint easily :) 

For example:

```python
# In "datasets/dataset_plugin.py", it is exposed publicly via "datasets/__init__.py"
Columns = Optional[Union[Iterable[str], str]]
```

Then in a custom project:

```python
from datasets import Columns
...
columns: Columns = [...]
@dataset(name=..., columns=columns)
```

Reply:
I'm not sure of the gains.  This adds more types to the interface, and the types would likely not be used, nor must be, because Python doesn't validate types until runtime...
This would be very useful in Scala, C# world...

Reply:
If the client uses `mypy` then they'd be able to run static analysis prior to runtime. Also even if we decide not to expose this publicly (which I still believe we should) at least if we used a type alias we'd be able to normalize down the types across it's usages internally to the library itself, currently it looks like the definition of the type of `columns` is duplicated across many files so having a canonical alias would at least benefit that use case that it'd be internally referenced everywhere similarly.

Reply:
Yeah I think there are several places within the code that'd be useful to normalize with type aliases. Typically the type alias will map to the parameter names as I'm not seeing any inconsistent parameter naming, just using type aliases would codify them and make them easier to reason about across the codebase in my opinion. There is some setup required for IDEs to allow easier understanding of what the underlying implementation of a type alias actually is but I'd say that's a secondary problem :)